### PR TITLE
MPS: add ranked tensors for addcmul ops instead of constants and update version_check

### PR DESCRIPTION
**Pull request by kulinseth**
Fixes #ISSUE_NUMBER

### CODE REVIEW

**Review by malfet — APPROVED** [review]
Looks good to me, can you please also remove unnecessary `@available` check from https://github.com/pytorch/pytorch/blob/5e03dfd36d5d5072e09b1ed1ccc7ee9b38b44bc7/aten/src/ATen/mps/EmptyTensor.cpp#L28

And perhaps refactor this check from the constructor into `hasCompatibleMPSFramework()` predicate?

File: aten/src/ATen/mps/MPSDevice.mm [hunk 1]
```diff
@@ -22,10 +22,17 @@
...
 MPSDevice::MPSDevice(): _mtl_device(nil) {
   // Check that MacOS 12.3+ version of MPS framework is available
-  id mpsCD = NSClassFromString(@"MPSGraphCompilationDescriptor");
-  if (![mpsCD instancesRespondToSelector:@selector(optimizationLevel)]) {
-    // According to https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraphcompilationdescriptor/3922624-optimizationlevel
-    // this means we are running on older MacOS
+  // Create the MPSGraph and check method introduced in 12.3+
+  // which is used by MPS backend.
+  id mpsCD = NSClassFromString(@"MPSGraph");
+  if (mpsCD == nil || ([mpsCD instancesRespondToSelector:@selector(LSTMWithSourceTensor:
```
Comment (malfet) [inline aten/src/ATen/mps/MPSDevice.mm:28]: Nit: one does not need an explicit `nil` check here, as `objc_msgSend` returns `NO` for empty objects
```suggestion
  if ([mpsCD instancesRespondToSelector:@selector(LSTMWithSourceTensor:
```

**Comment by kulinseth** [comment]
> hasCompatibleMPSFramework

```
 #if __is_target_os(macOS)
-  if (__builtin_available(macOS 12.3, *) || __builtin_available(macOSApplicationExtension 12.3, *)) {
+  if (at::hasMPS()) {
```
Does this work ?

**Comment by kulinseth** [comment]
@malfet updated with your comments. I tested it locally and test_mps is working fine on 12.3+ machine.

**Comment by suo** [comment]
Unfortunately had to revert this as it was stacked on top of https://github.com/pytorch/pytorch/pull/78230

**Comment by albanD** [comment]
cc @kulinseth make sure to re-add this one as well.

**Comment by malfet** [comment]
Well, actually this PR is a culprit for MacOS build breakages, as it contains a syntactic error, going to revert this and reland it soon

**Pull request closed without merging**